crates/topos-tce-api/src/runtime/sync_task.rs

```diff
@@ -0,0 +1,179 @@
+use crate::stream::StreamCommand;
+use futures::stream::FuturesUnordered;
+use std::collections::hash_map::Entry;
+use std::collections::HashMap;
+use std::future::{Future, IntoFuture};
+use std::pin::Pin;
+use tokio::sync::mpsc::Sender;
+use tokio_util::sync::CancellationToken;
+use topos_api::grpc::checkpoints::TargetStreamPosition;
+use topos_core::types::stream::CertificateTargetStreamPosition;
+use topos_core::types::CertificateDelivered;
+use topos_core::uci::SubnetId;
+use topos_tce_storage::{FetchCertificatesFilter, FetchCertificatesPosition, StorageClient};
+use tracing::{error, info};
+use uuid::Uuid;
+
+type TargetSubnetStreamPositions = HashMap<SubnetId, HashMap<SubnetId, TargetStreamPosition>>;
+pub(crate) type RunningTasks =
+    FuturesUnordered<Pin<Box<dyn Future<Output = SyncTaskStatus> + Send>>>;
+
+/// Status of a sync task
+///
+/// When registering a stream, a sync task is started to fetch certificates from the storage
+/// and push them to the stream.
+#[derive(Debug)]
+pub(crate) enum SyncTaskStatus {
+    ///  The sync task is active and started running
+    Running,
+    /// The sync task failed and reported an error
+    Error,
+    /// The sync task exited gracefully and is done pushing certificates to the stream
+    Done,
+    /// The sync task was cancelled externally
```

Comment:
For my education, can you elaborate on what this "external cancellation" is? The only place I've seen is when a sync task is replaced by another one (so that would be good to add here), but I'm not sure why/when this happens.